### PR TITLE
addded async log to keep high performance

### PR DESCRIPTION
**Pull request by ytwang0320**
added an independent thread to handle the logs.
need to add "$EscapeControlCharactersOnReceive off" in  "/etc/rsyslog.conf" if syslog is used
or there may be no line breaks since the logs are buffering in one.

### CODE REVIEW

File: src/log.c [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+    while (len--)
+    {
+        hash = hash * seed + (*str++);
+    }
+ 
+    return (hash & 0x7FFFFFFF);
+}
+
+
+static uint64_t log_get_time(char *time, int time_len)
```
Comment (AdamYYan) [inline src/log.c]: 如果时间的精度要求不高，能不能把time cache住？ 避免每次都做这些运算

File: src/log.c [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+static void log_signal_handler(int signum)
+{
+    if (signum == SIGABRT || signum == SIGSEGV) {
+        printf("\nSignal %d received, preparing to exit...\n",
+                signum);
+    }
+    log_slave_process();
+    log_buf_flush(rte_logs.file); 
+    signal(signum, SIG_DFL);
+    kill(getpid(), signum);
```
Comment (AdamYYan) [inline src/log.c]: 这里能直接同步等待log 落盘成功，然后使用信号的默认处理方式吗？ 需要额外调用一次kill吗？

File: src/sys_time.c [hunk 3]
```diff
@@ -60,7 +60,7 @@ void sys_start_time(void)
 
     time(&g_dpvs_timer);
     gettimeofday(&tv, &tz);
-    g_dpvs_timer += tz.tz_minuteswest*60;
+    g_dpvs_timer -= tz.tz_minuteswest*60;
```
[the remark this answers is not in the document]
Comment (ytwang0320) [reply src/sys_time.c]: 北京时间的tz.tz_minuteswest是负数（-480 ），这里需要加480得到正确的北京时间

File: include/log.h [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    int slow;
+    uint64_t slow_begin;
+    uint32_t missed;
+}log_stats;
+
+extern int g_dpvs_log_ready;
+
+#ifdef RTE_LOG
+#undef RTE_LOG
+#define RTE_LOG(l, t, ...)                  \
```
[the remark this answers is not in the document]
Comment (ytwang0320) [reply include/log.h]: 考虑直接替换就不影响调用的地方了

File: include/log.h [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    dpvs_log(RTE_LOG_ ## l,                   \
```
[the remark this answers is not in the document]
Comment (ytwang0320) [reply include/log.h]: 已添加g_dpvs_log_async_mode

File: src/log.c (hunk 1 above)
Comment (ytwang0320) [reply src/log.c]: 需要考虑时间在哪里update

File: include/log.h [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    uint64_t time;
+}log_buf_t;
+
+typedef struct log_stats{
+    int log_hash;
+    uint64_t log_begin;
+    int slow;
+    uint64_t slow_begin;
+    uint32_t missed;
+}log_stats;
```
Comment (ywc689) [inline include/log.h]:
```
typedef struct log_stats {
...
} log_stats_t
```

File: include/dpdk.h [hunk 7]
```diff
@@ -61,4 +61,14 @@ typedef uint8_t lcoreid_t;
 typedef uint16_t portid_t;
 typedef uint16_t queueid_t;
 
+
+#ifdef RTE_LOG
+extern int dpvs_log(uint32_t level, uint32_t logtype, const char *func, int line, const char *format, ...);
+#undef RTE_LOG
+#define RTE_LOG(l, t, ...)                  \
+    dpvs_log(RTE_LOG_ ## l,                   \
+        RTE_LOGTYPE_ ## t,  __func__, __LINE__, # t ": " __VA_ARGS__)
```
Comment (ywc689) [inline include/dpdk.h]: Is  the macro `RTE_LOG` duplicated with the one in "include/log.h"?

File: include/log.h [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    int log_len;
+    char data[0];           
+};
+
+typedef struct log_buf {
+    char buf[LOG_BUF_MAX_LEN];
+    int pos;
+    int level;
+    uint64_t time;
+}log_buf_t;
```
Comment (ywc689) [inline include/log.h]: `}log_buf_t;` -> `} log_buf_t;`
---------
Try to follow the same coding style with other codes  in DPVS. DPVS generally follow the [linux kernel code style](https://www.kernel.org/doc/html/v4.10/process/coding-style.html#).

File: src/log.c [hunk 9]
```diff
@@ -0,0 +1,361 @@
+#include <string.h>
+#include <syslog.h>
+#include <rte_mempool.h>
+#include "netif.h"
+#include "sys_time.h"
+#include "log.h"
+#include "common.h"
+#include "dpdk.h"
+#include "signal.h"
```
Comment (ywc689) [inline src/log.c]: include <signal.h>

File: src/log.c [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+    log_msg = rte_zmalloc("log_msg", sizeof(struct dpvs_log) + len, RTE_CACHE_LINE_SIZE);
+    if (unlikely(NULL == log_msg))
+        return NULL; 
+#endif
+    log_msg->log_level = level;
+    log_msg->log_type = type;
+    log_msg->cid = cid;
+    log_msg->log_len = len;
+    if (len)
+        memcpy(log_msg->data, data, len);
```
Comment (ywc689) [inline src/log.c]: use `rte_memcpy`

File: src/log.c [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+        }
+
+        cid = rte_lcore_id();
+
+        if (log_stats_info[cid].slow) {
+            /* set log limit rate to 5 sec and keep for 10 mins */
+            if (rte_get_timer_cycles() - log_stats_info[cid].slow_begin > LOG_SLOW_INTERNAL_TIME*rte_get_timer_hz()) {
+                log_stats_info[cid].slow = 0;
+            }
+            if ((rte_get_timer_cycles() - log_stats_info[cid].log_begin) < log_internal*rte_get_timer_hz()) {
```
Comment (ywc689) [inline src/log.c]: `log_internal * rte_get_timer_hz`

File: src/log.c [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+        if (logtype != RTE_LOGTYPE_USER1) {
+            rte_vlog(level, logtype, format, ap);
+            break;
+        }
+
+        cid = rte_lcore_id();
+
+        if (log_stats_info[cid].slow) {
+            /* set log limit rate to 5 sec and keep for 10 mins */
+            if (rte_get_timer_cycles() - log_stats_info[cid].slow_begin > LOG_SLOW_INTERNAL_TIME*rte_get_timer_hz()) {
```
Comment (ywc689) [inline src/log.c]: `LOG_SLOW_INTERNAL_TIME*rte_get_timer_hz()`
Keep the same spacing style for math operations for each side of comparsion.

File: src/log.c [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+        dpvs_async_log(level, logtype, cid, log_buf, len, off);
+    }while(0);
+    
+    va_end(ap);
+    return 0;
+}
+
+static int log_buf_flush(FILE *f)
+{
+    if (f == NULL ) {
```
Comment (ywc689) [inline src/log.c]: (f == NULL)

File: src/log.c [hunk 14]
```diff
@@ -0,0 +1,361 @@
...
+    return 0;
+}
+
+static int log_buf_timeout_flush(FILE *f, int timeout)
+{
+    uint64_t now;
+    
+    now = rte_get_timer_cycles();
+    
+    if (w_buf.pos && ((now - w_buf.time) >= timeout*rte_get_timer_hz())) {
```
Comment (ywc689) [inline src/log.c]: timeout * rte_get_timer_hz

File: src/log.c [hunk 15]
```diff
@@ -0,0 +1,361 @@
...
+{
+    struct dpvs_log *msg_log;
+    int ret = EDPVS_OK;
+    FILE *f = rte_logs.file;
+
+    /* dequeue LOG from ring, no lock for ring and w_buf */
+    while (0 == rte_ring_dequeue(log_ring, (void **)&msg_log)) { 
+        if (msg_log->log_len > LOG_BUF_MAX_LEN) {
+            dpvs_log_free(msg_log);
+            continue;
```
Comment (ywc689) [inline src/log.c]: Sliently drop the log message here? Can we not enqueue the oversized log message and return failure when sent?

File: src/log.c [hunk 16]
```diff
@@ -0,0 +1,361 @@
...
+
+static void log_signal_handler(int signum)
+{
+    if (signum == SIGABRT || signum == SIGSEGV) {
+        printf("\nSignal %d received, preparing to exit...\n",
+                signum);
+    }
+    log_slave_process();
+    log_buf_flush(rte_logs.file); 
+    signal(signum, SIG_DFL);
```
Comment (ywc689) [inline src/log.c]: DPVS config file uses SIGHUP signal. Does the SIG_DFL affect the config reload ?

File: src/log.c [hunk 17]
```diff
@@ -0,0 +1,361 @@
...
+    dp_vs_log_pool = rte_mempool_create(log_pool_name,
+                                log_pool_size,
+                                sizeof(struct dpvs_log) + DPVS_LOG_MAX_LINE_LEN,
+                                log_pool_cache,
+                                0, NULL, NULL, NULL, NULL,
+                                0, 0);
+    if (!dp_vs_log_pool) {
+        return EDPVS_DPDKAPIFAIL;
+    }
+#endif
```
Comment (ywc689) [inline src/log.c]: 1. Can we use memory pool in default? 
2. `CONFIG_DPVS_LOG_POOL_DEBUG` is a little confusing, `CONFIG_DPVS_LOG_MEM_POOL` is enough, i think.

**Pull request merged**